File: TGAPEngine/src/Material.cpp

#include "Material.h"

namespace engine{

Material::Material(const char * vertex_file_path,const char * fragment_file_path)
{
    std::cout<<"madWorldBase"<<std::endl;
    loadShaders(vertex_file_path,fragment_file_path);

	glUseProgram(this->program);

	this->cameraToClipUnif = glGetUniformLocation(program, "cameraToClip");
	this->modelToWorldUnif = glGetUniformLocation(program, "modelToWorld");
	this->worldToCameraUnif = glGetUniformLocation(program, "worldToCamera");
	for (int i = 0; i < 4; i++){
		std::ostringstream dir;
		dir << "lightDir[" << i << "]";
		std::ostringstream col;
		col << "lightCol[" << i << "]";
		std::ostringstream poi;
		poi << "pointLight[" << i << "]";
		this->lightDirUnif[i] = glGetUniformLocation(program, dir.str().c_str());
		this->lightColUnif[i] = glGetUniformLocation(program, col.str().c_str());
		this->pointLightUnif[i] = glGetUniformLocation(program, poi.str().c_str());
	}
	
	
	this->lightsNumberUnif = glGetUniformLocation(program, "lightsNumber");
}
Material::~Material()
{
    //dtor
}

void Material::loadShaders(const char * vertex_file_path,const char * fragment_file_path){

    // Create the shaders
    GLuint VertexShaderID = glCreateShader(GL_VERTEX_SHADER);
    GLuint FragmentShaderID = glCreateShader(GL_FRAGMENT_SHADER);

    // Read the Vertex Shader code from the file
    std::string VertexShaderCode;
    std::ifstream VertexShaderStream(vertex_file_path, std::ios::in);
    if(VertexShaderStream.is_open())
    {
        std::string Line = "";
        while(getline(VertexShaderStream, Line))
            VertexShaderCode += "\n" + Line;
        VertexShaderStream.close();
    }

    // Read the Fragment Shader code from the file
    std::string FragmentShaderCode;
    std::ifstream FragmentShaderStream(fragment_file_path, std::ios::in);
    if(FragmentShaderStream.is_open()){
        std::string Line = "";
        while(getline(FragmentShaderStream, Line))
            FragmentShaderCode += "\n" + Line;
        FragmentShaderStream.close();
    }

    GLint Result = GL_FALSE;
    int InfoLogLength;

    // Compile Vertex Shader
    printf("Compiling shader : %s\n", vertex_file_path);
    char const * VertexSourcePointer = VertexShaderCode.c_str();
    glShaderSource(VertexShaderID, 1, &VertexSourcePointer , NULL);
    glCompileShader(VertexShaderID);

    // Check Vertex Shader
    glGetShaderiv(VertexShaderID, GL_COMPILE_STATUS, &Result);
    glGetShaderiv(VertexShaderID, GL_INFO_LOG_LENGTH, &InfoLogLength);
    std::vector<char> VertexShaderErrorMessage(InfoLogLength);
	if (InfoLogLength > 0){
		glGetShaderInfoLog(VertexShaderID, InfoLogLength, NULL, &VertexShaderErrorMessage[0]);
		fprintf(stdout, "%s\n", &VertexShaderErrorMessage[0]);
	}
    

    // Compile Fragment Shader
    printf("Compiling shader : %s\n", fragment_file_path);
    char const * FragmentSourcePointer = FragmentShaderCode.c_str();
    glShaderSource(FragmentShaderID, 1, &FragmentSourcePointer , NULL);
    glCompileShader(FragmentShaderID);

    // Check Fragment Shader
    glGetShaderiv(FragmentShaderID, GL_COMPILE_STATUS, &Result);
    glGetShaderiv(FragmentShaderID, GL_INFO_LOG_LENGTH, &InfoLogLength);
    std::vector<char> FragmentShaderErrorMessage(InfoLogLength);
	if (InfoLogLength > 0){
		glGetShaderInfoLog(FragmentShaderID, InfoLogLength, NULL, &FragmentShaderErrorMessage[0]);
		fprintf(stdout, "%s\n", &FragmentShaderErrorMessage[0]);
	}
    

    // Link the program
    fprintf(stdout, "Linking program\n");
    GLuint ProgramID = glCreateProgram();
    glAttachShader(ProgramID, VertexShaderID);
    glAttachShader(ProgramID, FragmentShaderID);
    glBindFragDataLocation(ProgramID,0,"colorOut");
    glLinkProgram(ProgramID);

    // Check the program
    glGetProgramiv(ProgramID, GL_LINK_STATUS, &Result);
    glGetProgramiv(ProgramID, GL_INFO_LOG_LENGTH, &InfoLogLength);
    std::vector<char> ProgramErrorMessage( std::max(InfoLogLength, int(1)) );
    glGetProgramInfoLog(ProgramID, InfoLogLength, NULL, &ProgramErrorMessage[0]);
    fprintf(stdout, "%s\n", &ProgramErrorMessage[0]);

    glDeleteShader(VertexShaderID);
    glDeleteShader(FragmentShaderID);

    this->program = ProgramID;
}
void Material::use()
{
    glUseProgram(this->program);
}

void Material::setUniformCameraToClip(glm::mat4 matrix)
{
	glUniformMatrix4fv(this->cameraToClipUnif, 1, GL_FALSE, &matrix[0][0]);
}

void Material::setUniformModelToWorld(glm::mat4 matrix)
{
	glUniformMatrix4fv(this->modelToWorldUnif, 1, GL_FALSE, &matrix[0][0]);
}

void Material::setUniformWorldToCamera(glm::mat4 matrix)
{
	glUniformMatrix4fv(this->worldToCameraUnif, 1, GL_FALSE, &matrix[0][0]);
}
void Material::setUniformLightDir(glm::vec4 lightDir,int index)
{
	glUniform4fv(this->lightDirUnif[index], 1, &lightDir[0]);
}

void Material::setUniformLightCol(glm::vec3 lightCol, int index)
{
	glUniform3fv(this->lightColUnif[index], 1, &lightCol[0]);
}

void Material::setUniformPointLight(bool isPointLight,int index){
	if (isPointLight){
		glUniform1ui(this->pointLightUnif[index], 1);
	}
	else
	{
		glUniform1ui(this->pointLightUnif[index], 0);
	}
	
}
void Material::setUniformLightsNumber(unsigned int number){
	
	glUniform1ui(this->lightsNumberUnif, number);
	

}
}

File: CS/TGAPE_CS/Engine/Renderer/Material.cs
using CSGL;
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Linq;
using System.Numerics;
using System.Runtime.InteropServices;
using System.Text;
using System.Threading.Tasks;
using TGAPE_CS.Engine.Math;

namespace TGAPE_CS.Engine.Renderer
{
    public class Material
    {
        protected uint _shaderProgramHandle;
        protected int _cameraToClipSpaceUniformHandle;
        protected int _worldToCameraSpaceUniformHandle;
        protected int _modelToWorldSpaceUniformHandle;
        protected int[] _lightDirectionUniformHandles;
        protected int[] _lightColorUniformHandles;
        protected int[] _pointLightUniformHandles;
        protected int _lightsCountUniformHandle;

        public Material(string vertexFilePath, string fragmentFilePath)
        {
            _lightDirectionUniformHandles = new int[4];
            _lightColorUniformHandles = new int[4];
            _pointLightUniformHandles = new int[4];

            LoadShaders(vertexFilePath, fragmentFilePath);
            OpenGL.glUseProgram(_shaderProgramHandle);

            _cameraToClipSpaceUniformHandle = OpenGL.glGetUniformLocation(_shaderProgramHandle, "cameraToClip");//TODO:do not hardcode shit like this
            _modelToWorldSpaceUniformHandle = OpenGL.glGetUniformLocation(_shaderProgramHandle, "modelToWorld");//TODO:do not hardcode shit like this
            _worldToCameraSpaceUniformHandle = OpenGL.glGetUniformLocation(_shaderProgramHandle, "worldToCamera");//TODO:do not hardcode shit like this
            for(int i = 0; i<4; i++)//TODO:add dynamic lights - not max 4 lol
            {
                _lightDirectionUniformHandles[i] = OpenGL.glGetUniformLocation(_shaderProgramHandle, string.Format("lightDir[{0}]", i));
                _lightColorUniformHandles[i] = OpenGL.glGetUniformLocation(_shaderProgramHandle, string.Format("lightCol[{0}]", i));
                _pointLightUniformHandles[i] = OpenGL.glGetUniformLocation(_shaderProgramHandle, string.Format("pointLight[{0}]", i));
            }
            _lightsCountUniformHandle = OpenGL.glGetUniformLocation(_shaderProgramHandle, "lightsNumber");
        }

        public void LoadShaders(string vertexFilePath, string fragmentFilePath)
        {
            if(!File.Exists(vertexFilePath) || !File.Exists(fragmentFilePath))
            {
                //TODO:throw?
                return;
            }
            var vertexShaderHandle = OpenGL.glCreateShader(OpenGL.GL_VERTEX_SHADER);
            var fragmentShaderHandle = OpenGL.glCreateShader(OpenGL.GL_FRAGMENT_SHADER);
            var vertexShaderCode = File.ReadAllBytes(vertexFilePath);
            var fragmentShaderCode = File.ReadAllBytes(fragmentFilePath);

            int result = OpenGL.GL_FALSE;
            int infoLogLength = 0;
            
            Console.WriteLine(string.Format("Compiling shader : {0}\n"), vertexFilePath);
            int lenghtOfVertexCode = vertexShaderCode.Length * sizeof(byte);
            IntPtr vetrexCodePointer = Marshal.AllocHGlobal(lenghtOfVertexCode);
            Marshal.Copy(vertexShaderCode, 0, vetrexCodePointer, vertexShaderCode.Length);
            OpenGL.glShaderSource(vertexShaderHandle, 1, ref vetrexCodePointer, ref lenghtOfVertexCode);
            Marshal.FreeHGlobal(vetrexCodePointer);
            OpenGL.glCompileShader(vertexShaderHandle);

            OpenGL.glGetShaderiv(vertexShaderHandle, OpenGL.GL_COMPILE_STATUS, ref result);
            OpenGL.glGetShaderiv(vertexShaderHandle, OpenGL.GL_INFO_LOG_LENGTH, ref infoLogLength);
            if(infoLogLength > 0)
            {
                int lengthOfTheBuffer = 0;
                IntPtr logBuffer = Marshal.AllocHGlobal(infoLogLength);
                OpenGL.glGetShaderInfoLog(vertexShaderHandle, infoLogLength, ref lengthOfTheBuffer, logBuffer);

                byte[] buffer = new byte[lengthOfTheBuffer];
                Marshal.Copy(logBuffer, buffer, 0, lengthOfTheBuffer);
                Marshal.FreeHGlobal(logBuffer);
                var logMessage = Encoding.ASCII.GetString(buffer);
                Console.WriteLine(logMessage);
                if (result != OpenGL.GL_TRUE)
                {
                    Debugger.Break();
                    //TODO:THROW?
                    return;
                }
            }

            result = OpenGL.GL_FALSE;
            infoLogLength = 0;
            Console.WriteLine(string.Format("Compiling shader : {0}\n"), fragmentFilePath);
            int lenghtOfFragmentCode = fragmentShaderCode.Length * sizeof(byte);
            IntPtr fragmentCodePointer = Marshal.AllocHGlobal(lenghtOfFragmentCode);
            Marshal.Copy(fragmentShaderCode, 0, fragmentCodePointer, lenghtOfFragmentCode);
            OpenGL.glShaderSource(fragmentShaderHandle, 1, ref fragmentCodePointer, ref lenghtOfFragmentCode);
            Marshal.FreeHGlobal(fragmentCodePointer);
            OpenGL.glCompileShader(fragmentShaderHandle);

            OpenGL.glGetShaderiv(fragmentShaderHandle, OpenGL.GL_COMPILE_STATUS, ref result);
            OpenGL.glGetShaderiv(fragmentShaderHandle, OpenGL.GL_INFO_LOG_LENGTH, ref infoLogLength);
            if (infoLogLength > 0)
            {
                int lengthOfTheBuffer = 0;
                IntPtr logBuffer = Marshal.AllocHGlobal(infoLogLength);
                OpenGL.glGetShaderInfoLog(fragmentShaderHandle, infoLogLength, ref lengthOfTheBuffer, logBuffer);

                byte[] buffer = new byte[lengthOfTheBuffer];
                Marshal.Copy(logBuffer, buffer, 0, lengthOfTheBuffer);
                Marshal.FreeHGlobal(logBuffer);
                var logMessage = Encoding.ASCII.GetString(buffer);
                Console.WriteLine(logMessage);
                if (result != OpenGL.GL_TRUE)
                {
                    Debugger.Break();
                    //TODO:THROW?
                    return;
                }
            }

            result = OpenGL.GL_FALSE;
            infoLogLength = 0;
            Console.WriteLine("Linking program\n");
            uint programHandle = OpenGL.glCreateProgram();
            OpenGL.glAttachShader(programHandle, vertexShaderHandle);
            OpenGL.glAttachShader(programHandle, fragmentShaderHandle);
            OpenGL.glBindFragDataLocation(programHandle, 0, "colorOut");//TODO:this could be a metadata thingie not hardcoded
            OpenGL.glLinkProgram(programHandle);

            OpenGL.glGetProgramiv(programHandle, OpenGL.GL_LINK_STATUS, ref result);
            OpenGL.glGetProgramiv(programHandle, OpenGL.GL_INFO_LOG_LENGTH, ref infoLogLength);
            if (infoLogLength > 0)
            {
                int lengthOfTheBuffer = 0;
                IntPtr logBuffer = Marshal.AllocHGlobal(infoLogLength);
                OpenGL.glGetProgramInfoLog(programHandle, infoLogLength, ref lengthOfTheBuffer, logBuffer);

                byte[] buffer = new byte[lengthOfTheBuffer];
                Marshal.Copy(logBuffer, buffer, 0, lengthOfTheBuffer);
                Marshal.FreeHGlobal(logBuffer);
                var logMessage = Encoding.ASCII.GetString(buffer);
                Console.WriteLine(logMessage);
                if (result != OpenGL.GL_TRUE)
                {
                    Debugger.Break();
                    //TODO:THROW?
                    return;
                }
            }

            OpenGL.glDeleteShader(vertexShaderHandle);
            OpenGL.glDeleteShader(fragmentShaderHandle);

            _shaderProgramHandle = programHandle;
        }
        public void Use()
        {
            OpenGL.glUseProgram(_shaderProgramHandle);
        }
        public void SetUniformModelToWorld(Matrix4x4 modelToWorld)
        {
            var matrixAsArray = MatrixHelper.ConvertToAnArray(modelToWorld);
            OpenGL.glUniformMatrix4fv(_modelToWorldSpaceUniformHandle, 1, OpenGL.GL_TRUE, ref matrixAsArray[0]);
        }
        public void SetUniformCameraToClip(Matrix4x4 cameraToClip)
        {
            var matrixAsArray = MatrixHelper.ConvertToAnArray(cameraToClip);
            OpenGL.glUniformMatrix4fv(_cameraToClipSpaceUniformHandle, 1, OpenGL.GL_TRUE, ref matrixAsArray[0]);
        }
        public void SetUniformWorldToCamera(Matrix4x4 worldToCamera)
        {
            var matrixAsArray = MatrixHelper.ConvertToAnArray(worldToCamera);
            OpenGL.glUniformMatrix4fv(_worldToCameraSpaceUniformHandle, 1, OpenGL.GL_TRUE, ref matrixAsArray[0]);
        }
        public void SetUniformLightDirection(Vector4 lightDirection, int lightIndex)
        {
            float[] vectorAsArray = new float[4];
            lightDirection.CopyTo(vectorAsArray);
            OpenGL.glUniform4fv(_lightDirectionUniformHandles[lightIndex], 1, ref vectorAsArray[0]);
        }
        public void SetUniformLightColor(Vector3 lightColor, int lightIndex)
        {
            float[] vectorAsArray = new float[3];
            lightColor.CopyTo(vectorAsArray);
            OpenGL.glUniform3fv(_lightColorUniformHandles[lightIndex], 1, ref vectorAsArray[0]);
        }
        public void SetUniformPointLight(bool isAPointLight, int lightIndex)
        {
            OpenGL.glUniform1ui(_pointLightUniformHandles[lightIndex], isAPointLight ? 1u : 0u);
        }
        public void SetUniformLightsCount(uint count)
        {
            OpenGL.glUniform1ui(_lightsCountUniformHandle, count);
        }

    }
}
